### PR TITLE
Simplifying TPC reco workflow processors

### PR DESCRIPTION
**Pull request by matthiasrichter**
- publishing all TPC sectors in one time slice, even on the same lane
- updating ClustererSpec, ClusterDecoderRawSpec, and CATrackerSpec
  - supporting multiple parts per input route by using InputRecord iterator
    interface
  - drop the custom EOS handling, this is now handled by the DPL driver

### CODE REVIEW

**Comment by matthiasrichter** [comment]
Keeping this open as WIP for a couple of days, working on a extended InputRecord iterator API supporting filtering and multiple parts.

**Comment by davidrohr** [comment]
@matthiasrichter : Thx for simplifying this! Please make sure to test --ca-clusterer and/or --disable-mc options, to check all combinations.

File: Detectors/TPC/workflow/src/PublisherSpec.cxx [hunk 1]
```diff
@@ -150,34 +141,12 @@ DataProcessorSpec createPublisherSpec(PublisherConf const& config, bool propagat
...
-          // in digitizer workflow, -1 -> end of data, -2 noop
-          if (lane == 0) {
-            operation = -1;
-          }
-          o2::header::DataHeader::SubSpecificationType subSpec = processAttributes->outputIds[lane];
-          o2::tpc::TPCSectorHeader header{operation};
-          pc.outputs().snapshot(OutputRef{"output", subSpec, {header}}, subSpec);
-          if (propagateMC) {
-            pc.outputs().snapshot(OutputRef{"outputMC", subSpec, {header}}, subSpec);
-          }
```
[the remark this answers is not in the document]
Comment (matthiasrichter) [reply Detectors/TPC/workflow/src/PublisherSpec.cxx]: This was the custom end-of-stream-handling of the TPC as implemented originally in the TPC digitization. For the reco workflow this is now replaced by the DPL EOS handling. I'm going to remove the respective code from the other processors as well as soon as I have adjusted the digitizer workflow.

File: Detectors/TPC/workflow/src/PublisherSpec.cxx [hunk 2]
```diff
@@ -120,25 +120,16 @@ DataProcessorSpec createPublisherSpec(PublisherConf const& config, bool propagat
...
-    auto processingFct = [processAttributes, index = std::make_shared<unsigned>(0), propagateMC](ProcessingContext& pc) {
+    auto processingFct = [processAttributes, propagateMC](ProcessingContext& pc) {
       if (processAttributes->finished) {
         return;
       }
 
-      auto publish = [&processAttributes, &index, &pc, propagateMC]() {
-        auto& sectors = processAttributes->sectors;
+      bool eos = false;
+      auto& sectors = processAttributes->sectors;
```
[the remark this answers is not in the document]
Comment (matthiasrichter) [reply Detectors/TPC/workflow/src/PublisherSpec.cxx]: ok

File: Detectors/TPC/workflow/src/CATrackerSpec.cxx [hunk 3]
```diff
@@ -270,82 +270,90 @@ DataProcessorSpec getCATrackerSpec(bool processMC, bool caClusterer, std::vector
...
+              continue;
+            }
+
+            auto const* sectorHeader = DataRefUtils::getHeader<o2::tpc::TPCSectorHeader*>(ref);
+            if (sectorHeader == nullptr) {
+              // FIXME: think about error policy
+              LOG(ERROR) << "sector header missing on header stack";
+              return;
+            }
+            const int& sector = sectorHeader->sector;
```
[the remark this answers is not in the document]
Comment (matthiasrichter) [reply Detectors/TPC/workflow/src/CATrackerSpec.cxx]: no, changed

File: Detectors/TPC/workflow/src/CATrackerSpec.cxx [hunk 4]
```diff
@@ -270,82 +270,90 @@ DataProcessorSpec getCATrackerSpec(bool processMC, bool caClusterer, std::vector
...
+              !DataRefUtils::match(ref, {"check", ConcreteDataTypeMatcher{gDataOriginTPC, "CLUSTERNATIVE"}})) {
+            continue;
+          }
           auto const* sectorHeader = DataRefUtils::getHeader<o2::tpc::TPCSectorHeader*>(ref);
           if (sectorHeader == nullptr) {
             // FIXME: think about error policy
             LOG(ERROR) << "sector header missing on header stack";
             return;
           }
           const int& sector = sectorHeader->sector;
```
[the remark this answers is not in the document]
Comment (matthiasrichter) [reply Detectors/TPC/workflow/src/CATrackerSpec.cxx]: no, changed

**Comment by matthiasrichter** [comment]
Thanks for the comments, code is updated accordingly. From my manual check, the output of the workflow is the same with and without the PR. There is a slight difference in the number of found tracks with and without CA clusterer, consistently with and without PR.

I have no means to check the MC labels, however workflow runs with and without MC processing in all cases. @davidrohr, @shahor02, how about implementing a checker processor? If you can give me a macro how to check the tracking result against MC info, I will give it a try. Think we urgently need this.

**Comment by davidrohr** [comment]
@matthiasrichter : I have already ported part of the QA to O2. Can you try to add "--tracker-options 'qa'" to the workflow. It will create a couple of pdfs in the folder plots. Some of them are still not correctly filled, like the z resolution. But you can compare the output before and after. So far, only the results without --ca-clusterer are reasonable. There was a bug in the mc processing of the ca-clusterer, which was fixed this week, but I didn't check so far whether the QA looks ok now.

**Comment by matthiasrichter** [comment]
@ktf, any idea why there are two Travis builds now? I have checked, both are using the same build configuration, one has two blanks in the beginning of each line of the configuration file.

**Comment by matthiasrichter** [comment]
@davidrohr, somehow I'm missing the `o2sim_Kine.root` file, so the qa check does not work. And the tracker process hangs after the error message.

@sawenzel, @shahor02 where should `o2sim_Kine.root` come from, I have been running `sim_challenge.sh` to produce my test data.

**Comment by davidrohr** [comment]
@matthiasrichter : That is weird to me. I don't have an o2sim_Kine.root, but the tpc-reco-workflow works for me with the --tracker-options "qa" setting.

**Comment by shahor02** [comment]
@matthiasrichter: are you used data generated a few days ago? There was a large scale change a few days ago from @sawenzel, separating hits and kinematics files.

**Comment by matthiasrichter** [comment]
ok, so maybe something got out of sync there. The data set has been generated two or three weeks ago. I'm going to produce a new one.

**Comment by matthiasrichter** [comment]
the second Travis build is stuck, and from the configuration file it seems to be a duplication. Merging.

**Comment by davidrohr** [comment]
For reference, I just retried with the current dev.
Indeed, there is now an o2sim_Kine.root file, but the QA is still working.
Probably dataset and software need to be in sync, as you just said.

**Comment by matthiasrichter** [comment]
Yes, its working with a data set created with current dev (apart from #3233 ), and the plots are created.

**Comment by ktf** [comment]
@matthiasrichter I think it was just a travis fluke. I've seen a few issues in the last few day. Hopefully they are now solved.

**Pull request merged**